### ISSUE TITLE
Bug : when validate port_list

### ISSUE THREAD

**Issue by kholisrag** [mattjhayes/nmeta#115]
Bug when running...
It's because the validate function in port_list

```
Registered VCS backend: git
Registered VCS backend: hg
Registered VCS backend: svn
Registered VCS backend: bzr
loading app /home/pi/nmeta/nmeta/nmeta.py
loading app ryu.controller.ofp_handler
instantiating app ryu.controller.ofp_handler of OFPHandler
instantiating app /home/pi/nmeta/nmeta/nmeta.py of NMeta
03:52:34.007 config[10782] __init__ DEBUG dir_default=config dir_user=config/user config_filename=config.yaml
03:52:34.033 config[10782] ingest_config_file INFO Ingesting config file=/home/pi/nmeta/nmeta/config/config.yaml
03:52:34.131 config[10782] ingest_config_user INFO User-defined config file does not exist, file=/home/pi/nmeta/nmeta/config/user/config.yaml, skipping
03:52:34 config[10782] inherit_logging INFO Config logging now fully configured
03:52:34 nmeta[10782] __init__ INFO sys.path=['/usr/local/bin', '/usr/lib/python2.7', '/usr/lib/python2.7/plat-arm-linux-gnueabihf', '/usr/lib/python2.7/lib-tk', '/usr/lib/python2.7/lib-old', '/usr/lib/python2.7/lib-dynload', '/home/pi/.local/lib/python2.7/site-packages', '/usr/local/lib/python2.7/dist-packages', '/usr/lib/python2.7/dist-packages', '/usr/lib/python2.7/dist-packages/gtk-2.0', '/usr/lib/pymodules/python2.7', '/home/pi/nmeta/nmeta']
03:52:34 policy[10782] __init__ INFO User policy file=/home/pi/nmeta/nmeta/config/user/main_policy.yaml not found
03:52:34 policy[10782] __init__ INFO Opening default policy file=/home/pi/nmeta/nmeta/config/main_policy.yaml
03:52:34 policy[10782] validate CRITICAL Voluptuous detected a problem where=port_set, exception=expected int for dictionary value @ data['port_list'][0]['DPID']
Exiting nmeta. Please fix error in main_policy.yaml
```
When I commented the code in line 842 is work normal again,

**Comment by mattjhayes** [comment]
Sorry, on holiday sitting on a beach in Fiji. I'll have a look in a few days. BTW, suggest not changing the supplied main_policy.yaml and instead creating your own one in the user sub directory. It will fully replace the supplied one if it exists. Are you able to supply the policy file you're using please?

**Comment by kholisrag** [comment]
happy and good luck for your holiday... 
I create main_policy long time ago, and testing in my laptop as controller... it's work...
now I implement nmeta to raspberry pi 3 to act as controller...
it's working when I test and add qos_treatment now... without that line of validation

btw, this is my main_policy

```
---
#*** Main Policy for nmeta - Static Regression Test
#*** Written in YAML
#
tc_rules:
    # Traffic Classification Rulesets and Rules
    tc_ruleset_1:
        # Video Rulesets
        - comment: Video Rulesets
          match_type: any
          conditions_list:
              - match_type: any
                classifiers_list:
                    - identity_service_dns_re: '.*\.youtube\.com'
                    - identity_service_dns_re: '.*\.youtube\*'
                    - identity_service_dns_re: '.*\.googlevideo\.com'
                    - identity_service_dns_re: '.*\.vimeo\.com'
                    - identity_service_dns_re: '.*\.vidio\.com'
          actions:
              set_desc: "Video Rulesets"
              qos_treatment: high_priority

        # HTTP/HTTPS Rulesets
        - comment: HTTP/HTTPS and Social Media Rulesets
          match_type: any
          conditions_list:
              - match_type: any
                classifiers_list:
                  - identity_service_dns_re: '.*\.google\*'
                  - identity_service_dns_re: '.*\.facebook\.com'
                  - identity_service_dns_re: '.*\.fbcdn\.*'
                  - identity_service_dns_re: '.*\.twitter\.com'
                  - identity_service_dns_re: '.*\.google\.com'
                  - tcp_dst: 80
                  - tcp_dst: 443
          actions:
              set_desc: "HTTP/HTTPS and Social Media Rulesets"
              qos_treatment: default_priority

        # OpenFlow Protocol Traffic Rulesets
        - comment: OpenFlow Protocol Traffic
          match_type: any
          conditions_list:
              - match_type: any
                classifiers_list:
                    - tcp_src: 6633
                    - tcp_dst: 6633
          actions:
              set_desc: "OpenFlow Protocol Traffic"
              qos_treatment: high_priority
        # Voice Rulesets
        - comment: Voice Rulesets (Whatsapp)
          match_type: any
          conditions_list:
              - match_type: any
                classifiers_list:
                    - identity_service_dns_re: '.*\.whatsapp\.com'
                    - identity_service_dns_re: '.*\.whatsapp\.net'
                    - identity_service_dns_re: '.*\.whatsapp\.*'
                    - tcp_src: 5060
                    - tcp_dst: 5060
                    - tcp_src: 5064
                    - tcp_dst: 5064
          actions:
              set_desc: "Voice Rulesets (Whatsapp)"
              qos_treatment: default_priority

        # Except All Rulesets
        - comment: Except All Rulesets
          match_type: any
          conditions_list:
              - match_type: any
                classifiers_list:
                  - location_src: internal
          actions:
              set_desc: "Except All Rulesets"
              qos_treatment: constrained_bw

#
qos_treatment:
    # Control Quality of Service (QoS) treatment mapping of
    #  names to output queue numbers:
    low_priority: 0
    default_priority: 1
    high_priority: 2
    constrained_bw: 99
#
port_sets:
    # Port Sets control what data plane ports policies and
    #  features are applied on. Names must be unique.
    port_set_list:
        - name: port_set_location_internal
          port_list:
              - name: Mikrobits-internal
                DPID: 277468410886552
                ports: 3-10
                vlan_id: 0
              - name: Tlwr1043ndv4-internal
                DPID: 168082750879242
                ports: 7,10-13
                vlan_id: 0
        - name: port_set_location_external
          port_list:
              - name: Mikrobits-external
                DPID: 277468410886552
                ports: 3
                vlan_id: 0
              - name: Tlwr1043ndv4-external
                DPID: 168082750879242
                ports: 8,9
                vlan_id: 0
#
locations:
    # Locations are logical groupings of ports. Takes first match.
    locations_list:
        - name: internal
          port_set_list:
            - port_set: port_set_location_internal
        - name: external
          port_set_list:
            - port_set: port_set_location_external
    default_match: unknown

```

**Comment by mattjhayes** [comment]
Thanks for the policy file, looks like you're trying some interesting stuff. Is this for research, teaching or hobby?

I'm having trouble reproducing the bug relating to verifying the DPID in a port_list. The only way I can reproduce it is by entering a non-integer for the DPID, which fails verification but is expected behaviour. When I run your policy it verifies fine. Any ideas?

**Comment by kholisrag** [comment]
I am working on my final project for my bachelor degree... implementation bandwidth management in sdn network... now I interested to learn python... I use your code for learning bit by a bit...

sigh... I sugget it's about the repository in raspbian, it's quiet left behind ubuntu... that's not critical problem I think... because is work normal when I use without validate port_list...

**Comment by kholisrag** [comment]
@mattjhayes how to fix about this?
is there somthing wrong in my main_policy.yaml
tc_static[2184] is_match_ip_space ERROR error=E1000021 Exception converting to IPAddress object. Exception <class 'netaddr.core.AddrFormatError'>, failed to detect a valid IP address from '', <traceback object at 0x74c33f80>

**Comment by mattjhayes** [comment]
I've just seen same error in my physical lab, where the controller is Ubuntu 14.04. Am upgrading to see if it's 14.04 specific. Will advise

**Comment by mattjhayes** [comment]
I've updated develop branch with improved error message - let me know what it shows, and what version of linux you're running. Thanks.

**Comment by kholisrag** [comment]
this my hostnamectl info:
   Static hostname: raspicontroller
         Icon name: computer
           Chassis: n/a
        Machine ID: 4207055ce1b14c12b59787680b4c7651
           Boot ID: 11c36124e25340af881132c47d1a8100
  Operating System: Raspbian GNU/Linux 8 (jessie)
            Kernel: Linux 4.9.24-v7+
      Architecture: arm

tc_static[2184] is_match_ip_space ERROR error=E1000021 Exception converting to IPAddress object. Exception <class 'netaddr.core.AddrFormatError'>, failed to detect a valid IP address from '', <traceback object at 0x74c33f80>
SOLVED with add `location` in `classifiers_list`

**Comment by kholisrag** [comment]
this is the traceback...
`21:09:50 tc_static[17550] is_match_ip_space ERROR error=E1000021 Exception converting  to IPAddress object. Exception <class 'netaddr.core.AddrFormatError'>, failed to detect a valid IP address from '', ['  File "/home/kholisrag/nmeta/nmeta/tc_static.py", line 350, in is_match_ip_space\n    ip_addr_object = IPAddress(ip_addr)\n', '  File "/home/kholisrag/.local/lib/python2.7/site-packages/netaddr/ip/__init__.py", line 306, in __init__\n    \'address from %r\' % addr)\n']`

**Comment by kholisrag** [comment]
This line
```
        #*** Convert the IP address to a netaddr IPAddress object:
        try:
            ip_addr_object = IPAddress(ip_addr)
        except:
            exc_type, exc_value, exc_traceback = sys.exc_info()
            self.logger.error("error=E1000021 "
                            "Exception converting %s to IPAddress object. "
                            "Exception %s, %s, %s",
                            ip_addr, exc_type, exc_value,
                            traceback.format_tb(exc_traceback))
            return 0
```
And this line
```
                if self._module is None:
                    raise AddrFormatError('failed to detect a valid IP ' \
                        'address from %r' % addr)
```

**Comment by mattjhayes** [comment]
Thanks for revised output, it shows ip_addr is empty, but not sure of cause yet. Will advise.

**Comment by mattjhayes** [comment]
I've found the cause, it occurs when a non-IP packet is checked against a static IP classifier. I will update code to fix, but in the interim should be safe to ignore these errors.

**Comment by mattjhayes** [comment]
I've put a fix in for this bug, and released code to master branch as 0.4.0
Let me know if this issue can be closed please. Thanks.
